devito/tools/utils.py

```diff
@@ -156,10 +157,31 @@ def filter_ordered(elements, key=None):
     """
     seen = set()
     if key is None:
-        key = lambda x: x
+        try:
+            unordered, inds = np.unique(elements, return_index=True)
+            return unordered[np.argsort(inds)].tolist()
+        except:
+            return sorted(list(set(elements)), key=elements.index)
```

Comment:
no, this is wrong. `sorted` may change the input ordering. This would only be OK for `filter_sorted`

Reply:
No this is correct because of  `key=elements.index`, that's the only reason sorted is called to revert `set`  to original ordering

Reply:
right. I think I'm an idiot. Ignore me. 

How slower/faster this fallback is?

Reply:
ah, but there's one thing you wanna do: `except ...` Missing the exception type

Reply:
>except ... Missing the exception type
Kinda left it out on purpose because it depends on what type of input there is. Can be sorting errors if non-sortables (relational error), or type errors or ...
Basically it try numpy if you can, if not go back to list defaults, can put a generic one though

Reply:
>How slower/faster this fallback is?

See test, ~ 2 times better on smallish, for the index part get around 10k better for very large shot recs